pkg/testworkflows/testworkflowprocessor/action/optimize.go

```diff
@@ -0,0 +1,324 @@
+		}
+		return expressions.MustCompile(refReplacements[name]), true
+	})
+	wrapEndRef := expressions.NewMachine().RegisterAccessor(func(name string) (interface{}, bool) {
+		if result, ok := refResults[name]; ok {
+			return expressions.MustCompile(result), true
+		}
+		return nil, false
+	})
+	for i := range actions {
+		if actions[i].CurrentStatus != nil {
+			actions[i].CurrentStatus = common.Ptr(simplifyExpression(*actions[i].CurrentStatus, wrapStartRef))
+			actions[i].CurrentStatus = common.Ptr(simplifyExpression(*actions[i].CurrentStatus, wrapEndRef))
+		}
+		if actions[i].Declare != nil {
+			actions[i].Declare.Condition = simplifyExpression(actions[i].Declare.Condition, wrapStartRef)
+			actions[i].Declare.Condition = simplifyExpression(actions[i].Declare.Condition, wrapEndRef)
+		}
+		if actions[i].Result != nil {
+			actions[i].Result.Value = simplifyExpression(actions[i].Result.Value, wrapStartRef)
+			actions[i].Result.Value = simplifyExpression(actions[i].Result.Value, wrapEndRef)
+		}
+	}
+
+	// Detect immediately skipped steps
+	skipped := make(map[string]struct{})
+	for i := range actions {
+		if actions[i].Declare != nil {
+			v, err := expressions.EvalExpressionPartial(actions[i].Declare.Condition)
+			if err == nil && v.Static() != nil {
+				b, err := v.Static().BoolValue()
+				if err == nil && !b {
+					skipped[actions[i].Declare.Ref] = struct{}{}
+				}
+			}
+		}
+	}
+
+	// List all the results
+	results := make(map[string]expressions.Expression)
+	conditions := make(map[string]expressions.Expression)
+	for i := range actions {
+		if actions[i].Result != nil {
+			var err error
+			refs[actions[i].Result.Ref] = struct{}{}
+			results[actions[i].Result.Ref], err = expressions.EvalExpressionPartial(actions[i].Result.Value)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if actions[i].Declare != nil {
+			var err error
+			conditions[actions[i].Declare.Ref], err = expressions.EvalExpressionPartial(actions[i].Declare.Condition)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if actions[i].Execute != nil {
+			refs[actions[i].Execute.Ref] = struct{}{}
+		}
+	}
+
+	// Pre-resolve conditions
+	currentStatus := expressions.MustCompile("true")
+	executed := make(map[string]struct{})
+	for i := range actions {
+		// Update current status
+		if actions[i].CurrentStatus != nil {
+			var err error
+			currentStatus, err = expressions.Compile(*actions[i].CurrentStatus)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Mark step as executed
+		if actions[i].Execute != nil {
+			executed[actions[i].Execute.Ref] = struct{}{}
+		} else if actions[i].End != nil {
+			executed[*actions[i].End] = struct{}{}
+		}
+
+		// Simplify the condition
+		if actions[i].Declare != nil {
+			// TODO: Handle `never` and other aliases
+			machine := expressions.NewMachine().RegisterAccessor(func(name string) (interface{}, bool) {
+				if name == "passed" || name == "success" {
```

Comment:
are these const?

Reply:
It rather shouldn't - `passed` and `success` are just built-in variables, and it could be any kind of expression